internal/agent/runtime/docker.go

```diff
@@ -0,0 +1,191 @@
+package runtime
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"regexp"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/client"
+	"github.com/go-logr/logr"
+	"github.com/tinkerbell/tink/internal/agent"
+	"github.com/tinkerbell/tink/internal/agent/runtime/internal"
+	"github.com/tinkerbell/tink/internal/agent/workflow"
+	"github.com/tinkerbell/tink/internal/ptr"
+)
+
+var _ agent.ContainerRuntime = &Docker{}
+
+// Docker is a docker runtime that satisfies agent.ContainerRuntime.
+type Docker struct {
+	log    logr.Logger
+	client *client.Client
+}
+
+// Run satisfies agent.ContainerRuntime.
+func (d *Docker) Run(ctx context.Context, a workflow.Action) error {
+	// We need the image to be available before we can create a container.
+	image, err := d.client.ImagePull(ctx, a.Image, types.ImagePullOptions{})
+	if err != nil {
+		return fmt.Errorf("docker: %w", err)
+	}
+	defer image.Close()
+
+	// Docker requires everything to be read from the images ReadCloser for the image to actually
+	// be pulled. We may want to log image pulls in a circular buffer somewhere for debugability.
+	if _, err = io.Copy(io.Discard, image); err != nil {
+		return fmt.Errorf("docker: %w", err)
+	}
+	// Close should be idempotent and we don't need the handle beyond this point.
+	image.Close()
+
+	// TODO: Support all the other things on the action such as volumes.
+	cfg := container.Config{
+		Image: a.Image,
+		Env:   toDockerEnv(a.Env),
+	}
+
+	failureFiles, err := internal.NewFailureFiles()
+	if err != nil {
+		return fmt.Errorf("create action failure files: %w", err)
+	}
+	defer failureFiles.Close()
+
+	hostCfg := container.HostConfig{
+		Mounts: []mount.Mount{
+			{
+				Type:   mount.TypeBind,
+				Source: failureFiles.ReasonPath(),
+				Target: ReasonMountPath,
+			},
+			{
+				Type:   mount.TypeBind,
+				Source: failureFiles.MessagePath(),
+				Target: MessageMountPath,
+			},
+		},
+	}
+
+	containerName := toContainerName(a.ID)
+
+	// Docker uses the entrypoint as the default command. The Tink Action Cmd property is modeled
+	// as being the command launched in the container hence it is used as the entrypoint. Args
+	// on the action are therefore the command portion in Docker.
+	if a.Cmd != "" {
+		cfg.Entrypoint = append(cfg.Entrypoint, a.Cmd)
+	}
+	if len(a.Args) > 0 {
+		cfg.Cmd = append(cfg.Cmd, a.Args...)
+	}
+
+	// TODO: Figure out container logging. We probably want to save it somewhere for debugability.
+
+	create, err := d.client.ContainerCreate(ctx, &cfg, &hostCfg, nil, nil, containerName)
+	if err != nil {
+		return fmt.Errorf("docker: %w", err)
+	}
+
+	// Always try to remove the container on exit.
+	defer func() {
+		// We can't use the context passed to Run() as it may have been cancelled.
+		err := d.client.ContainerRemove(context.Background(), create.ID, types.ContainerRemoveOptions{})
```

Comment:
should we enable all the `ContainerRemoveOptions`?

Reply:
```
type ContainerRemoveOptions struct {
	RemoveVolumes bool
	RemoveLinks   bool
	Force         bool
}
```

On Mac with Docker Desktop I've seen force leave objects lying on disk that couldn't be cleaned up by a following `docker system prune --volumes` (I had to manually remove files). I don't know how widespread that impact is, or how challenging it would be to figure out, so I'm not sure if we'd want that enabled, particularly given this is all in-memory and goes away on restart. Given action names are unique I don't foresee any conflicts with respect to launching more actions.

I don't think we want to remove volumes because we allow users to share data between actions via volumes. 

I don't see any issues with link removal.

Reply:
On action name uniqueness, I've realized [the action ID is unique within the context of a workflow](https://github.com/tinkerbell/tink/blob/main/api/v1alpha2/workflow.go#L57), but if a second workflow executes there could be a collision. I think we could fix that by committing to UUIDs for action IDs.

Reply:
thanks, Chris.

You're right we probably shouldn't remove volumes.

Currently, the main use case for the agent is to run in Hook. The status-quo is to use [`Force: true`](https://github.com/tinkerbell/tink/blob/cc1f8f13d92299fde6bb46a90298c720ee742711/cmd/tink-worker/worker/container_manager.go#L145), so we know it works in Hook. I'm concerned about resource constrained machines where memory and disk space in Hook are an issue.

Reply:
OK, lets stick with the status quo and force delete. The volumes will likely be in-memory and cleaned on restart anyway.

Reply:
Should be addressed.